ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,698 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Consensus.GSM (tests) where
+
+import           Control.Concurrent.Class.MonadSTM.Strict.TVar.Checked
+import           Control.Monad.Class.MonadAsync (withAsync)
+import           Control.Monad.Class.MonadFork (MonadFork, yield)
+import           Control.Monad.Class.MonadSTM
+import qualified Control.Monad.Class.MonadTime.SI as SI
+import qualified Control.Monad.Class.MonadTimer.SI as SI
+import           Control.Monad.IOSim (IOSim, runSimOrThrow)
+import           Data.Kind (Type)
+import           Data.List ((\\))
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import           Data.Time (diffTimeToPicoseconds)
+import qualified Data.TreeDiff as TD
+import           GHC.Generics (Generic, Generic1)
+import qualified Ouroboros.Consensus.Node.GSM as GSM
+import           Ouroboros.Network.PeerSelection.LedgerPeers.Type (LedgerStateJudgement (..))
+import           Ouroboros.Consensus.Util.IOLike (IOLike)
+import           Test.QuickCheck (choose, elements, oneof, shrink)
+import qualified Test.QuickCheck as QC
+import qualified Test.QuickCheck.Monadic as QC
+import           Test.StateMachine (Concrete, Symbolic)
+import           Test.Consensus.IOSimQSM.Test.StateMachine.Sequential (runCommands')
+import qualified Test.StateMachine as QSM
+import qualified Test.StateMachine.Types as QSM
+import qualified Test.StateMachine.Types.Rank2 as QSM
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestEnv (adjustQuickCheckTests)
+import           Test.Util.ToExpr ()
+import           Test.Tasty (TestTree)
+import           Test.Tasty.QuickCheck (testProperty)
+
+tests :: [TestTree]
+tests = [
+    adjustQuickCheckTests (* 100) $ testProperty "GSM" prop_sequential
+  ]
+
+-----
+
+-- The following definitions are in the exact same order as the QSM tutorial at
+-- <https://github.com/stevana/quickcheck-state-machine/blob/3748220bffacb61847e35e3808403f3e7187a7c6/README.md>.
+
+-- | TODO restarts (or maybe just killing the GSM thread)
+type Command :: (Type -> Type) -> Type
+data Command r =
+    Disconnect UpstreamPeer
+    -- ^ INVARIANT must be an existing peer
+    --
+    -- Mocks the necessary ChainSync client behavior.
+  |
+    ExtendSelection S
+    -- ^ INVARIANT 'selectionIsBehind'
+    --
+    -- NOTE Harmless to assume it only advances by @'B' 1@ at a time.
+  |
+    ModifyCandidate UpstreamPeer B
+    -- ^ INVARIANT existing peer
+    --
+    -- Mocks the necessary ChainSync client behavior.
+  |
+    NewCandidate UpstreamPeer B
+    -- ^ INVARIANT new peer
+    --
+    -- Mocks the necessary ChainSync client behavior.
+  |
+    ReadJudgment
+  |
+    ReadMarker
+  |
+    StartIdling UpstreamPeer
+    -- ^ INVARIANT existing peer, not idling
+  |
+    TimePasses Int
+    -- ^ tenths of a second
```

Comment:
- I want a bare `Int` so that the counterexamples are more legible and ergonamic.
- I want to use 1 second slot duration, but I want the time passes to be more granular than that.
- But not _too_ granular, for the sake of legibility again.

That's the motivation for the current use of bare "tenths".